Fix build errors for builds for non-standard build locations

lbhm opened the pull request:
While preparing for the addition of [WebP support for DALI](https://github.com/NVIDIA/DALI/issues/3203), I wanted to build the DALI deps from source. Being on a machine without root access, I choose a PREFIX inside my home directory and encountered several errors. Since this repo does not have an issues section, this PR is semi-PR and semi-discussion.

Explanation of my changes:
- `build_protobuf` and `build_cmake` had hardcoded paths so I changed them.
- `build_opencv` and `build_openjpeg` use `mkdir build` which will crash if the folder already exists. This was kinda annoying during my debugging so I changed it to `mkdir -p. Does this have any negative side effects you are aware of?
- The `lmdb` Makefile is somewhat confusing to me because it uses both `DESTDIR` and `prefix` in the `install target`. Maybe my knowledge of Makefiles is too limited but in my opinion this is redundant. In any way, since the `lmdb` repo does not use CMake or autoconf, I had to prepend `prefix` on the `make install` call again. `DESTDIR=""` is a hack I chose because of the redundancy. Maybe a patch would be a cleaner solution here? I wasn't sure how to create a proper patch file here though.
- I added `WITH_WEBP=ON` to `build_opencv` in prepration for [DALI#3203](https://github.com/NVIDIA/DALI/issues/3203). I only added this before posting a PR to the main DALI repo because according to the [OpenCV docs](https://docs.opencv.org/4.5.2/db/d05/tutorial_config_reference.html), `WITH_WEBP=ON` is the default setting so making this explicit shouldn't actually change anything.

On a side note: Building FFmpeg with `build_ffmpeg` using the `nasm` assembler did not work on my machine (AMD Epyc 7742, Ubuntu 20.04 LTS) unless I disabled the MMX instruction set. `yasm` on the other hand worked fine so maybe it would be helpful to add a note to the DALI build from source guide to use `yasm` over `nasm` for FFmpeg.

lbhm commented:
As an additional question: Are any of the libraries supposed to link against each other during the build process? I would assume yes since some of the libs use CMake. In that case, a non-standard `INSTALL_PREFIX` would also have to be passed as an additional library location to CMake, autoconf, or Make (depending on the lib), right?
The CMake installation location wasn't an issue in my case because I already have a system-wide CMake installation on the machine but I suppose some of the other libraries could be built incorrectly if they do not find the previously built libraries.

JanuszL commented on build_scripts/build_protobuf.sh:
```diff
@@ -31,7 +31,7 @@ if [ "${CC_COMP}" != "gcc" ]; then
       ${BUILD_ARCH_OPTION} \
       ${SYSROOT_ARG} \
-      --with-protoc=/usr/local/bin/protoc \
+      --with-protoc=${INSTALL_PREFIX}/bin/protoc \
```
I think it won't work when building with JetPack. In such case the `INSTALL_PREFIX` is the directory with binaries for target architecture, while the `protoc` should be build for host platform.

lbhm replied:
Fixed in 3030b7f3b127f281bca2bc91fb37207a4afcd2a6

JanuszL commented on build_scripts/build_cmake.sh:
```diff
@@ -16,7 +16,7 @@
 # CMake
 pushd third_party/CMake
-./bootstrap --parallel=$(grep ^processor /proc/cpuinfo | wc -l)
+./bootstrap --parallel=$(grep ^processor /proc/cpuinfo | wc -l) --prefix=${INSTALL_PREFIX}
```
The same problem as with protobuf applies - it should be installed in host, not target architecture destination directory.
Maybe there should be one more variable - `HOST_INSTALL_PREFIX`?

lbhm replied:
Fixed in 3030b7f3b127f281bca2bc91fb37207a4afcd2a6

JanuszL commented:
Hi @lbhm,
> * `build_protobuf` and `build_cmake` had hardcoded paths so I changed them.

Indeed it should be changed, but also it should account for cross-compilation cases - like JetPack. See my comment in the review.

> * `build_opencv` and `build_openjpeg` use `mkdir build` which will crash if the folder already exists. This was kinda annoying during my debugging so I changed it to `mkdir -p. Does this have any negative side effects you are aware of?

It should be fine. Good catch.

> * The `lmdb` Makefile is somewhat confusing to me because it uses both `DESTDIR` and `prefix` in the `install target`. Maybe my knowledge of Makefiles is too limited but in my opinion this is redundant. In any way, since the `lmdb` repo does not use CMake or autoconf, I had to prepend `prefix` on the `make install` call again. `DESTDIR=""` is a hack I chose because of the redundancy. Maybe a patch would be a cleaner solution here? I wasn't sure how to create a proper patch file here though.

I think you can update https://github.com/NVIDIA/DALI_deps/blob/main/patches/Makefile-lmdb.patch, Jut pull the repo, do changes and create a patch file from the diff.

> * I added `WITH_WEBP=ON` to `build_opencv` in prepration for [DALI#3203](https://github.com/NVIDIA/DALI/issues/3203). I only added this before posting a PR to the main DALI repo because according to the [OpenCV docs](https://docs.opencv.org/4.5.2/db/d05/tutorial_config_reference.html), `WITH_WEBP=ON` is the default setting so making this explicit shouldn't actually change anything.

👍 

> On a side note: Building FFmpeg with `build_ffmpeg` using the `nasm` assembler did not work on my machine (AMD Epyc 7742, Ubuntu 20.04 LTS) unless I disabled the MMX instruction set. `yasm` on the other hand worked fine so maybe it would be helpful to add a note to the DALI build from source guide to use `yasm` over `nasm` for FFmpeg.

In my case (Intel) nasm works. I guess it is an issue specific to FFmpeg itself, not DALI per se, so I have mixed filling about adding a note to the DALI docs about it.

lbhm commented:
Hi @JanuszL,

> Indeed it should be changed, but also it should account for cross-compilation cases - like JetPack.

I see. I'm not familiar with the cross-compilation setup so would you think adding a `HOST_INSTALL_PREFIX` will suffice? Will you discuss this team-internally or how should we proceed?

> I think you can update https://github.com/NVIDIA/DALI_deps/blob/main/patches/Makefile-lmdb.patch, Jut pull the repo, do changes and create a patch file from the diff.

Would you agree that just removing `DESTDIR` from the `install` target should be a fix since `prefix` is already part of `libdir`, `includedir`, etc.? If yes, I can create an according patch file.

JanuszL commented:
Hi @lbhm,

HOST_INSTALL_PREFIX should do. I can help with testing.

> Would you agree that just removing DESTDIR from the install target should be a fix since prefix is already part of libdir, includedir, etc.? If yes, I can create an according patch file.

I think it should do as well.

lbhm commented:
@JanuszL: I added a `HOST_INSTALL_PREFIX` which by default also defines the `INSTALL_PREFIX`. I also updated the patch file for the lmdb Makefile.
It would be great if you could test the changes I propose. I do not have docker or root access on the machine I currently use for testing so I cannot cover all of the cases.

JanuszL commented:
@lbhm,

Looks good. Let me trigger a CI job for all your PRs. You should see the status in https://github.com/NVIDIA/DALI/pull/3206.

lbhm commented:
Good morning @JanuszL,
the build failed but I'm not sure if I can see why from the outside. Do you investigate this internally or is there anything I can do to help?

JanuszL commented on build_scripts/build_protobuf.sh:
```diff
@@ -17,7 +17,7 @@
 pushd third_party/protobuf
 ./autogen.sh
-./configure CXXFLAGS="-fPIC" --prefix=/usr/local --disable-shared 2>&1 > /dev/null
+./configure CXXFLAGS="-fPIC" --prefix=${INSTALL_PREFIX} --disable-shared 2>&1 > /dev/null
```
```suggestion
./configure CXXFLAGS="-fPIC" --prefix=${HOST_INSTALL_PREFIX} --disable-shared 2>&1 > /dev/null
```

JanuszL replied:
As we want to install protoc for host, and bellow we install protobuff for target just saying it should use protoc from host.

lbhm replied:
Makes sense. Fixed in f631458bfd78a835d26c635e3232de1b13176b54.

JanuszL commented:
Good morning @lbhm,

I'm sorry but we cannot expose our CI to the outside world due to several reasons. I know it is not great but that is what it is.
I checked the logs and it failed in the build for Jetston - cross-compilation case. I think it is because of changes `build_protobuf.sh`. Can you check my comment?

JanuszL reviewed (approved):
Let us see how the build works.
